GDT.md

```diff
@@ -0,0 +1,220 @@
+- DS: Data selector, where general memory access can happen.
+- SS: Stack selector, where push/pop operations can happen.
+- ES: Extra selector, intended for use with string operations, no specific purpose.
+- FS: F selector, no specific purpose. Sys V ABI uses it for thread local storage.
+- GS: G selector, no specific purpose. Sys V ABI uses it for process local storage, commonly used for cpu-local storage in kernels due to `swapgs` instruction.
+
+Address types:
+- Logical address: addresses the programmer deals with.
+- Linear address: logical address after translation through segmentation (logical_address + selector_base).
+- Physical address: linear address translated through paging, maps to an actual memory location in RAM.
+
+It's worth noting if segmentation is ignored, logical and linear addresses are the same. <br>
+If paging is disabled, linear and physical addresses are the same.
+
+## Segmentation
+
+Segmentation is a mechanism for separating regions of memory into code and data, to help secure operating systems and hardware against potential security issues, and simpifies running multiple processes.
+
+How it works is pretty simple, each GDT descriptor defines a 'segment' in memory, using a base address and limit. 
+When a descriptor is loaded into the appropriate segment register, it creates a window into memory with the specified permissions. All memory outside of this segment has no persmissions (read, write, execute) unless specified by another segment.
+
+The idea is to place code in one region of memory, and then create a descriptor with a base and limit that only expose that region of memory to the cpu. Any attempts to fetch instructions from outside that region will result in a #GP fault being triggered, and the kernel will intervene.
+
+Accessing memory inside a segment is done relative to it's base. Lets say you have a segment with a base of 0x1000,
+and some data in memory at address 0x1100.
+The data would be accessed at address 0x100 (assuming the segment is the active DS), as addressed are translated as `segment_base + offset`. In this case the segment base is 0x1000, and the offset is 0x100.
+
+Segments can also be explicitly referenced. To load something at offset 0x100 into the ES region, you can use `mov es:0x100, $rax`. This would perform the translation from logical address to linear address using ES instead of DS (the default for data), a common example is when an interrupt occurs while the cpu is in ring 3, it will switch to ring 0 and load the appropriate descriptors into the segment registers.
+
+### Segment Registers
+
+The various segment registers and their uses are outline below. There are some tricks to load a descriptor from the GDT into a segment register. They can't be mov'd into directly, so you'll need to use a scratch register to change their value. The cpu will also automatically reload segment registers on certain events (see the manual for these). 
```

Comment:
i'm not totally sure (you are the native speaker so maybe it's me being wrong) should be outlined below?

Reply:
you're correct on that one ;)